sopel/irc/backends.py

```diff
@@ -5,346 +5,330 @@
-                                       suppress_ragged_eofs=True,
-                                       cert_reqs=ssl.CERT_REQUIRED,
-                                       ca_certs=self.ca_certs)
-            # connect to host specified in config first
-            try:
-                ssl.match_hostname(self.ssl.getpeercert(), self.host)
-            except ssl.CertificateError:
-                # the host in config and certificate don't match
-                LOGGER.error("hostname mismatch between configuration and certificate")
-                # check (via exception) if a CNAME matches as a fallback
-                has_matched = False
-                for hostname in get_cnames(self.host):
-                    try:
-                        ssl.match_hostname(self.ssl.getpeercert(), hostname)
-                        LOGGER.warning(
-                            "using {0} instead of {1} for TLS connection"
-                            .format(hostname, self.host))
-                        has_matched = True
-                        break
-                    except ssl.CertificateError:
-                        pass
-
-                if not has_matched:
-                    # everything is broken
-                    LOGGER.error("Invalid certificate, no hostname matches.")
-                    # TODO: refactor access to bot's settings
-                    if hasattr(self.bot.settings.core, 'pid_file_path'):
-                        # TODO: refactor to quit properly (no "os._exit")
-                        os.unlink(self.bot.settings.core.pid_file_path)
-                        os._exit(1)
-        self.set_socket(self.ssl)
-        LOGGER.info('Connection accepted by the server...')
-        LOGGER.debug('Starting job scheduler for connection timeout...')
-        self.timeout_scheduler.start()
+            self.bot.on_message(data)
+
+        # cancel timeout tasks when reading loop ends
+        self._cancel_timeout_tasks()
+
+    # run & connection
+
+    def get_connection_kwargs(self) -> Dict:
+        """Return the keyword arguments required to initiate connection."""
+        ssl_context: Optional[ssl.SSLContext] = None
+
+        if self._use_ssl:
+            ssl_context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
+            if self._certfile is not None:
+                # load_cert_chain requires a certfile (cannot be None)
+                ssl_context.load_cert_chain(
+                    certfile=self._certfile,
+                    keyfile=self._keyfile,
+                )
+
+            if not self._verify_ssl:
+                # deactivate SSL verification for hostname & certificate
+                ssl_context.check_hostname = False
+                ssl_context.verify_mode = ssl.CERT_NONE
+            else:
+                # verify location if verify_mode is not CERT_NONE
+                ssl_context.load_verify_locations(self._ca_certs)
```

Comment:
This will raise `TypeError: cafile, capath and cadata cannot be all omitted` if `self._ca_certs=None`, so I don't think `Optional[]` is appropriate for `__init__.ca_certs` unless there's a check or the default is loaded here instead of core_section.py

Aside: Should we really be doing [`_find_certs()`](https://github.com/sopel-irc/sopel/blob/4322aa45d345327f27e1616af460937fa3e5e991/sopel/config/core_section.py#L26) instead of trusting Python to have a sane default set by upstream or distro packagers? This smells bad.

Reply:
I agree with you. Here is what I'll do and what I suggest to do:

* in this PR, I'll keep as much as the previous behavior as possible (i.e. I'll keep the whole `ca_certs` things for now)
* in a future PR, we should rework that specific section, and everything related to that

Removing old behaviors that comes from older version of Python is, in my opinion, a good thing. It's even better to do so regarding ssl, as I know it was kind of a nightmare to deal with back in Python 2.4/2.7.